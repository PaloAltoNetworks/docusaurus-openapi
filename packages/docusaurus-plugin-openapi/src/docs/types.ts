/* ============================================================================
 * Copyright (c) Cloud Annotations
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */

import type { VersionBanner } from "@docusaurus/plugin-content-docs";
import type { Tag, FrontMatterTag } from "@docusaurus/utils";
import type { ContentPaths } from "@docusaurus/utils/lib/markdownLinks";

export interface DocObject {}

export type DocFrontMatter = {
  // Front matter uses snake case
  id?: string;
  title?: string;
  tags?: FrontMatterTag[];
  hide_title?: boolean;
  hide_table_of_contents?: boolean;
  keywords?: string[];
  image?: string;
  description?: string;
  slug?: string;
  sidebar_label?: string;
  sidebar_position?: number;
  sidebar_class_name?: string;
  sidebar_custom_props?: Record<string, unknown>;
  displayed_sidebar?: string | null;
  pagination_label?: string;
  custom_edit_url?: string | null;
  parse_number_prefixes?: boolean;
  toc_min_heading_level?: number;
  toc_max_heading_level?: number;
  pagination_next?: string | null;
  pagination_prev?: string | null;
};

export type LastUpdateData = {
  lastUpdatedAt?: number;
  formattedLastUpdatedAt?: string;
  lastUpdatedBy?: string;
};

export type DocMetadataBase = LastUpdateData & {
  id: string; // TODO legacy versioned id => try to remove
  unversionedId: string; // TODO new unversioned id => try to rename to "id"
  version: string;
  title: string;
  description: string;
  source: string; // @site aliased posix source => "@site/docs/folder/subFolder/subSubFolder/myDoc.md"
  sourceDirName: string; // posix path relative to the versioned docs folder (can be ".") => "folder/subFolder/subSubFolder"
  slug: string;
  permalink: string;
  sidebarPosition?: number;
  editUrl?: string | null;
  tags: Tag[];
  frontMatter: DocFrontMatter & Record<string, unknown>;
};

export type VersionMetadata = ContentPaths & {
  versionName: string; // 1.0.0
  versionLabel: string; // Version 1.0.0
  versionPath: string; // /baseUrl/docs/1.0.0
  tagsPath: string;
  versionEditUrl?: string | undefined;
  versionEditUrlLocalized?: string | undefined;
  versionBanner: VersionBanner | null;
  versionBadge: boolean;
  versionClassName: string;
  isLast: boolean;
  sidebarFilePath: string | false | undefined; // versioned_sidebars/1.0.0.json
  routePriority: number | undefined; // -1 for the latest docs
};
