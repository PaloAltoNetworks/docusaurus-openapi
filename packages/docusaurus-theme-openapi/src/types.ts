/* ============================================================================
 * Copyright (c) Palo Alto Networks
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */

import type { JSONSchema4, JSONSchema6, JSONSchema7 } from "json-schema";

export interface ThemeConfig {
  api?: {
    proxy?: string;
    authPersistance?: false | "localStorage" | "sessionStorage";
  };
}

interface Map<T> {
  [key: string]: T;
}

export type JSONSchema = JSONSchema4 | JSONSchema6 | JSONSchema7;
export type SchemaObject = Omit<
  JSONSchema,
  | "type"
  | "allOf"
  | "oneOf"
  | "anyOf"
  | "not"
  | "items"
  | "properties"
  | "additionalProperties"
> & {
  // OpenAPI specific overrides
  type?: "string" | "number" | "integer" | "boolean" | "object" | "array";
  allOf?: SchemaObject[];
  oneOf?: SchemaObject[];
  anyOf?: SchemaObject[];
  not?: SchemaObject;
  items?: SchemaObject;
  properties?: Map<SchemaObject>;
  additionalProperties?: boolean | SchemaObject;

  // OpenAPI additions
  nullable?: boolean;
  discriminator?: DiscriminatorObject;
  readOnly?: boolean;
  writeOnly?: boolean;
  xml?: XMLObject;
  externalDocs?: ExternalDocumentationObject;
  example?: any;
  deprecated?: boolean;
};

export interface DiscriminatorObject {
  propertyName: string;
  mapping?: Map<string>;
}

export interface XMLObject {
  name?: string;
  namespace?: string;
  prefix?: string;
  attribute?: boolean;
  wrapped?: boolean;
}

export interface ExternalDocumentationObject {
  description?: string;
  url: string;
}
