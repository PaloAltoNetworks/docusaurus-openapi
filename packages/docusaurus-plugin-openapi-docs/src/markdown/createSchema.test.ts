/* ============================================================================
 * Copyright (c) Palo Alto Networks
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */

import * as prettier from "prettier";

import { createNodes } from "./createSchema";
import { SchemaObject } from "../openapi/types";

describe("createNodes", () => {
  describe("oneOf", () => {
    it("should create readable MODs for oneOf primitive properties", async () => {
      const schema: SchemaObject = {
        "x-tags": ["clown"],
        type: "object",
        properties: {
          oneOfProperty: {
            oneOf: [
              {
                type: "object",
                properties: {
                  noseLength: {
                    type: "number",
                  },
                },
                required: ["noseLength"],
                description: "Clown's nose length",
              },
              {
                type: "array",
                items: {
                  type: "string",
                },
                description: "Array of strings",
              },
              {
                type: "boolean",
              },
              {
                type: "number",
              },
              {
                type: "string",
              },
            ],
          },
        },
      };
      expect(
        await Promise.all(
          createNodes(schema, "request").map(
            async (md: any) => await prettier.format(md, { parser: "babel" })
          )
        )
      ).toMatchSnapshot();
    });
  });

  describe("allOf", () => {
    it("should render same-level properties with allOf", async () => {
      const schema: SchemaObject = {
        allOf: [
          {
            type: "object",
            properties: {
              allOfProp1: {
                type: "string",
              },
              allOfProp2: {
                type: "string",
              },
            },
          },
        ],
        properties: {
          parentProp1: {
            type: "string",
          },
          parentProp2: {
            type: "string",
          },
        },
      };

      expect(
        await Promise.all(
          createNodes(schema, "response").map(
            async (md: any) => await prettier.format(md, { parser: "babel" })
          )
        )
      ).toMatchSnapshot();
    });

    it("should correctly merge nested properties from multiple allOf schemas", async () => {
      const schema: SchemaObject = {
        allOf: [
          {
            type: "object",
            properties: {
              outerProp1: {
                type: "object",
                properties: {
                  innerProp1: {
                    type: "string",
                  },
                },
              },
            },
          },
          {
            type: "object",
            properties: {
              outerProp2: {
                type: "object",
                properties: {
                  innerProp2: {
                    type: "number",
                  },
                },
              },
            },
          },
        ],
      };

      expect(
        await Promise.all(
          createNodes(schema, "response").map(
            async (md: any) => await prettier.format(md, { parser: "babel" })
          )
        )
      ).toMatchSnapshot();
    });

    it("should correctly handle shared required properties across allOf schemas", async () => {
      const schema: SchemaObject = {
        allOf: [
          {
            type: "object",
            properties: {
              sharedProp: {
                type: "string",
              },
            },
            required: ["sharedProp"],
          },
          {
            type: "object",
            properties: {
              anotherProp: {
                type: "number",
              },
            },
            required: ["anotherProp"],
          },
        ],
      };

      expect(
        await Promise.all(
          createNodes(schema, "response").map(
            async (md: any) => await prettier.format(md, { parser: "babel" })
          )
        )
      ).toMatchSnapshot();
    });

    // Could not resolve values for path:"properties.conflictingProp.type". They are probably incompatible. Values:
    // "string"
    // "number"
    // eslint-disable-next-line jest/no-commented-out-tests
    // it("should handle conflicting properties in allOf schemas", async () => {
    //   const schema: SchemaObject = {
    //     allOf: [
    //       {
    //         type: "object",
    //         properties: {
    //           conflictingProp: {
    //             type: "string",
    //           },
    //         },
    //       },
    //       {
    //         type: "object",
    //         properties: {
    //           conflictingProp: {
    //             type: "number",
    //           },
    //         },
    //       },
    //     ],
    //   };

    //   expect(
    //     await Promise.all(
    //       createNodes(schema, "response").map(
    //         async (md: any) => await prettier.format(md, { parser: "babel" })
    //       )
    //     )
    //   ).toMatchSnapshot();
    // });

    // Could not resolve values for path:"type". They are probably incompatible. Values:
    // "object"
    // "array"
    // eslint-disable-next-line jest/no-commented-out-tests
    // it("should handle mixed data types in allOf schemas", async () => {
    //   const schema: SchemaObject = {
    //     allOf: [
    //       {
    //         type: "object",
    //         properties: {
    //           mixedTypeProp1: {
    //             type: "string",
    //           },
    //         },
    //       },
    //       {
    //         type: "array",
    //         items: {
    //           type: "number",
    //         },
    //       },
    //     ],
    //   };

    //   expect(
    //     await Promise.all(
    //       createNodes(schema, "response").map(
    //         async (md: any) => await prettier.format(md, { parser: "babel" })
    //       )
    //     )
    //   ).toMatchSnapshot();
    // });

    it("should correctly deep merge properties in allOf schemas", async () => {
      const schema: SchemaObject = {
        allOf: [
          {
            type: "object",
            properties: {
              deepProp: {
                type: "object",
                properties: {
                  innerProp1: {
                    type: "string",
                  },
                },
              },
            },
          },
          {
            type: "object",
            properties: {
              deepProp: {
                type: "object",
                properties: {
                  innerProp2: {
                    type: "number",
                  },
                },
              },
            },
          },
        ],
      };

      expect(
        await Promise.all(
          createNodes(schema, "response").map(
            async (md: any) => await prettier.format(md, { parser: "babel" })
          )
        )
      ).toMatchSnapshot();
    });

    // eslint-disable-next-line jest/no-commented-out-tests
    // it("should handle discriminator with allOf schemas", async () => {
    //   const schema: SchemaObject = {
    //     allOf: [
    //       {
    //         type: "object",
    //         discriminator: {
    //           propertyName: "type",
    //         },
    //         properties: {
    //           type: {
    //             type: "string",
    //           },
    //         },
    //       },
    //       {
    //         type: "object",
    //         properties: {
    //           specificProp: {
    //             type: "string",
    //           },
    //         },
    //       },
    //     ],
    //   };

    //   expect(
    //     await Promise.all(
    //       createNodes(schema, "response").map(
    //         async (md: any) => await prettier.format(md, { parser: "babel" })
    //       )
    //     )
    //   ).toMatchSnapshot();
    // });
  });

  describe("additionalProperties", () => {
    it.each([
      [
        {
          allOf: [
            {
              oneOf: [
                {
                  type: "object",
                  properties: {
                    type: {
                      type: "string",
                      enum: ["nose"],
                    },
                  },
                  required: ["type"],
                },
                {
                  type: "object",
                  properties: {
                    type: {
                      type: "string",
                      enum: ["mouth"],
                    },
                  },
                  required: ["type"],
                },
                {
                  type: "object",
                  properties: {
                    type: {
                      type: "string",
                      enum: ["eyes"],
                    },
                    default: {
                      type: "string",
                    },
                  },
                  required: ["type"],
                },
              ],
            },
            {
              type: "object",
              properties: {
                description: {
                  type: "string",
                  description: "Description of the body part.",
                },
              },
              required: ["description"],
            },
          ],
        },
      ],
      [
        {
          type: "array",
          items: { type: "object", properties: { a: "string", b: "number" } },
        },
      ],
      [{ type: "string" }],
      [{ type: "number" }],
      [{ type: "integer" }],
      [{ type: "boolean" }],
      [false],
      [true],
      [{}],
    ] as [SchemaObject["additionalProperties"]][])(
      "should handle additionalProperties: %p",
      async (additionalProperties) => {
        const schema: SchemaObject = {
          type: "object",
          additionalProperties,
        };

        expect(
          await Promise.all(
            createNodes(schema, "request").map(
              async (md: any) => await prettier.format(md, { parser: "babel" })
            )
          )
        ).toMatchSnapshot();
      }
    );
  });
});
