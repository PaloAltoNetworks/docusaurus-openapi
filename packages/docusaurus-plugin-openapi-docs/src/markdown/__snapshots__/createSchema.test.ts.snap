// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createNodes additionalProperties should handle additionalProperties: {"allOf": [Array]} 1`] = `
Array [
  "<SchemaItem collapsible={true} className={\\"schemaItem\\"}>
  <details style={{}} className={\\"openapi-markdown__details\\"}>
    <summary style={{}}>
      <span className={\\"openapi-schema__container\\"}>
        <strong className={\\"openapi-schema__property\\"}>property name*</strong>
        <span className={\\"openapi-schema__name\\"}>object</span>
      </span>
    </summary>
    <div style={{ marginLeft: \\"1rem\\" }}>
      <div>
        <span className={\\"badge badge--info\\"} style={{ marginBottom: \\"1rem\\" }}>
          oneOf
        </span>
        <SchemaTabs>
          <TabItem label={\\"MOD1\\"} value={\\"0-item-properties\\"}>
            <SchemaItem
              collapsible={false}
              name={\\"type\\"}
              required={true}
              schemaName={\\"string\\"}
              qualifierMessage={\\"**Possible values:** [\`nose\`]\\"}
              schema={{ type: \\"string\\", enum: [\\"nose\\"] }}
            ></SchemaItem>
          </TabItem>
          <TabItem label={\\"MOD2\\"} value={\\"1-item-properties\\"}>
            <SchemaItem
              collapsible={false}
              name={\\"type\\"}
              required={true}
              schemaName={\\"string\\"}
              qualifierMessage={\\"**Possible values:** [\`mouth\`]\\"}
              schema={{ type: \\"string\\", enum: [\\"mouth\\"] }}
            ></SchemaItem>
          </TabItem>
          <TabItem label={\\"MOD3\\"} value={\\"2-item-properties\\"}>
            <SchemaItem
              collapsible={false}
              name={\\"type\\"}
              required={true}
              schemaName={\\"string\\"}
              qualifierMessage={\\"**Possible values:** [\`eyes\`]\\"}
              schema={{ type: \\"string\\", enum: [\\"eyes\\"] }}
            ></SchemaItem>
            <SchemaItem
              collapsible={false}
              name={\\"default\\"}
              required={false}
              schemaName={\\"string\\"}
              qualifierMessage={undefined}
              schema={{ type: \\"string\\" }}
            ></SchemaItem>
          </TabItem>
        </SchemaTabs>
      </div>
      <SchemaItem
        collapsible={false}
        name={\\"description\\"}
        required={true}
        schemaName={\\"string\\"}
        qualifierMessage={undefined}
        schema={{
          type: \\"string\\",
          description: \\"Description of the body part.\\",
        }}
      ></SchemaItem>
    </div>
  </details>
</SchemaItem>;
",
]
`;

exports[`createNodes additionalProperties should handle additionalProperties: {"items": [Object], "type": "array"} 1`] = `
Array [
  "<SchemaItem collapsible={true} className={\\"schemaItem\\"}>
  <details style={{}} className={\\"openapi-markdown__details\\"}>
    <summary style={{}}>
      <span className={\\"openapi-schema__container\\"}>
        <strong className={\\"openapi-schema__property\\"}>property name*</strong>
        <span className={\\"openapi-schema__name\\"}>object[]</span>
      </span>
    </summary>
    <div style={{ marginLeft: \\"1rem\\" }}>
      <li>
        <div
          style={{
            fontSize: \\"var(--ifm-code-font-size)\\",
            opacity: \\"0.6\\",
            marginLeft: \\"-.5rem\\",
            paddingBottom: \\".5rem\\",
          }}
        >
          Array [
        </div>
      </li>
      <SchemaItem
        collapsible={false}
        name={\\"a\\"}
        required={false}
        schemaName={\\"\\"}
        qualifierMessage={undefined}
        schema={\\"string\\"}
      ></SchemaItem>
      <SchemaItem
        collapsible={false}
        name={\\"b\\"}
        required={false}
        schemaName={\\"\\"}
        qualifierMessage={undefined}
        schema={\\"number\\"}
      ></SchemaItem>
      <li>
        <div
          style={{
            fontSize: \\"var(--ifm-code-font-size)\\",
            opacity: \\"0.6\\",
            marginLeft: \\"-.5rem\\",
          }}
        >
          ]
        </div>
      </li>
    </div>
  </details>
</SchemaItem>;
",
]
`;

exports[`createNodes additionalProperties should handle additionalProperties: {"type": "boolean"} 1`] = `
Array [
  "<SchemaItem
  name={\\"property name*\\"}
  required={false}
  schemaName={\\"boolean\\"}
  qualifierMessage={undefined}
  schema={{ type: \\"boolean\\" }}
  collapsible={false}
  discriminator={false}
></SchemaItem>;
",
]
`;

exports[`createNodes additionalProperties should handle additionalProperties: {"type": "integer"} 1`] = `
Array [
  "<SchemaItem
  name={\\"property name*\\"}
  required={false}
  schemaName={\\"integer\\"}
  qualifierMessage={undefined}
  schema={{ type: \\"integer\\" }}
  collapsible={false}
  discriminator={false}
></SchemaItem>;
",
]
`;

exports[`createNodes additionalProperties should handle additionalProperties: {"type": "number"} 1`] = `
Array [
  "<SchemaItem
  name={\\"property name*\\"}
  required={false}
  schemaName={\\"number\\"}
  qualifierMessage={undefined}
  schema={{ type: \\"number\\" }}
  collapsible={false}
  discriminator={false}
></SchemaItem>;
",
]
`;

exports[`createNodes additionalProperties should handle additionalProperties: {"type": "string"} 1`] = `
Array [
  "<SchemaItem
  name={\\"property name*\\"}
  required={false}
  schemaName={\\"string\\"}
  qualifierMessage={undefined}
  schema={{ type: \\"string\\" }}
  collapsible={false}
  discriminator={false}
></SchemaItem>;
",
]
`;

exports[`createNodes additionalProperties should handle additionalProperties: {} 1`] = `
Array [
  "<SchemaItem
  name={\\"property name*\\"}
  required={false}
  schemaName={\\"any\\"}
  qualifierMessage={undefined}
  schema={{ type: \\"object\\", additionalProperties: {} }}
  collapsible={false}
  discriminator={false}
></SchemaItem>;
",
]
`;

exports[`createNodes additionalProperties should handle additionalProperties: false 1`] = `Array []`;

exports[`createNodes additionalProperties should handle additionalProperties: true 1`] = `
Array [
  "<SchemaItem
  name={\\"property name*\\"}
  required={false}
  schemaName={\\"any\\"}
  qualifierMessage={undefined}
  schema={{ type: \\"object\\", additionalProperties: true }}
  collapsible={false}
  discriminator={false}
></SchemaItem>;
",
]
`;

exports[`createNodes should create readable MODs for oneOf primitive properties 1`] = `
Array [
  "<SchemaItem collapsible={true} className={\\"schemaItem\\"}>
  <details style={{}} className={\\"openapi-markdown__details\\"}>
    <summary style={{}}>
      <span className={\\"openapi-schema__container\\"}>
        <strong className={\\"openapi-schema__property\\"}>oneOfProperty</strong>
        <span className={\\"openapi-schema__name\\"}>object</span>
      </span>
    </summary>
    <div style={{ marginLeft: \\"1rem\\" }}>
      <div>
        <span className={\\"badge badge--info\\"} style={{ marginBottom: \\"1rem\\" }}>
          oneOf
        </span>
        <SchemaTabs>
          <TabItem label={\\"MOD1\\"} value={\\"0-item-properties\\"}>
            <div
              style={{ marginTop: \\".5rem\\", marginBottom: \\".5rem\\" }}
              className={\\"openapi-schema__summary\\"}
            >
              Clown's nose length
            </div>
            <SchemaItem
              collapsible={false}
              name={\\"noseLength\\"}
              required={true}
              schemaName={\\"number\\"}
              qualifierMessage={undefined}
              schema={{ type: \\"number\\" }}
            ></SchemaItem>
          </TabItem>
          <TabItem label={\\"MOD2\\"} value={\\"1-item-properties\\"}>
            <div
              style={{ marginTop: \\".5rem\\", marginBottom: \\".5rem\\" }}
              className={\\"openapi-schema__summary\\"}
            >
              Array of strings
            </div>
            <li>
              <div
                style={{
                  fontSize: \\"var(--ifm-code-font-size)\\",
                  opacity: \\"0.6\\",
                  marginLeft: \\"-.5rem\\",
                  paddingBottom: \\".5rem\\",
                }}
              >
                Array [
              </div>
            </li>
            <div style={{ marginTop: \\".5rem\\", marginBottom: \\".5rem\\" }}>
              string
            </div>
            <li>
              <div
                style={{
                  fontSize: \\"var(--ifm-code-font-size)\\",
                  opacity: \\"0.6\\",
                  marginLeft: \\"-.5rem\\",
                }}
              >
                ]
              </div>
            </li>
          </TabItem>
          <TabItem label={\\"MOD3\\"} value={\\"2-item-properties\\"}>
            <div style={{ marginTop: \\".5rem\\", marginBottom: \\".5rem\\" }}>
              boolean
            </div>
          </TabItem>
          <TabItem label={\\"MOD4\\"} value={\\"3-item-properties\\"}>
            <div style={{ marginTop: \\".5rem\\", marginBottom: \\".5rem\\" }}>
              number
            </div>
          </TabItem>
          <TabItem label={\\"MOD5\\"} value={\\"4-item-properties\\"}>
            <div style={{ marginTop: \\".5rem\\", marginBottom: \\".5rem\\" }}>
              string
            </div>
          </TabItem>
        </SchemaTabs>
      </div>
    </div>
  </details>
</SchemaItem>;
",
]
`;
